Fingerprint processor

ycombinator opened the pull request:
Resolves https://github.com/elastic/beats/issues/11173.

This PR implements a `fingerprint` processor, similar to [Logstash's `fingerprint` filter plugin](https://www.elastic.co/guide/en/logstash/current/plugins-filters-fingerprint.html).

The processor will take the following configuration options:

| Name | Required? | Default | Description |
| ----- | ----- | ----- | ----- |
| `fields` | Yes | | List of fields to use as the source of the fingerprint |
| `ignore_missing` | No | `false` | Whether to ignore missing fields |
| `target_field` | No | `fingerprint` | Field in which the computed fingerprint should be stored |
| `method` | No | `sha256` | Algorithm to use for computing the fingerprint. Must be one of: `md5`, `sha1`, `sha256` (default), `sha384`, `sha512`|
| `encoding`| No | `hex` | Encoding to use on the fingerprint value. Must be one of `hex` (default), `base32`, or `base64` |

andrewkroh commented:
This reminds me of https://github.com/andrewkroh/beats-processor-fingerprint. As well as https://github.com/elastic/beats/pull/1872.

ycombinator commented:
> This reminds me of https://github.com/andrewkroh/beats-processor-fingerprint. As well as https://github.com/elastic/beats/pull/1872.

Thanks for flagging this down, @andrewkroh; I wasn't aware.

@urso Looks like there's prior art (the processor in Andrew's repo) as well as discussions going on in the ES team (see https://github.com/elastic/elasticsearch/issues/16938, which is linked off #1872). Does it still make sense for me to continue working on this PR?

urso commented:
> Does it still make sense for me to continue working on this PR?

I think yes, it makes sense.

The one by Andrew is a private one. Its not to be found in Beats. Maybe we can take some of it.

there is also the Logstash fingerprint processor. I hope the ES one will be similar to the Logstash one. But I didn't check yet.

ycombinator commented:
Per the discussion in https://github.com/elastic/elasticsearch/pull/47047#issuecomment-545980835, we are going to resume working on a `fingerprint` processor in Beats.

@andrewkroh Since you've already built a fingerprint processor in your repo, did you want to put up a PR with that code to the `beats` repo? If not, I'll continue working on this PR while looking at your work.

andrewkroh commented:
> did you want to put up a PR with that code to the beats repo? If not, I'll continue working on this PR while looking at your work.

No, but please copy anything you want from my repo. 👍

ycombinator commented:
jenkins, test this

urso commented on libbeat/processors/fingerprint/config.go:
```diff
@@ -0,0 +1,34 @@
+	Fields      []string       `config:"fields" validate:"required"` // Source fields to compute fingerprint from
+	TargetField string         `config:"target_field"`               // Target field for the fingerprint
+	Encoding    encodingMethod `config:"encoding"`                   // Encoding to use for target field value
+}
```
what if a field is missing? We want some 'ignore_missing' setting?

ycombinator replied:
Sounds good. Will add.

urso commented on libbeat/processors/fingerprint/fingerprint.go:
```diff
@@ -0,0 +1,134 @@
+	}
+	str += "|"
+	return str, nil
+}
```
The hash types in the stdlib implement io.Writer. Instead of allocating memory on every call consider streaming the fields into into an writer.

like:

```
func writeKeys(to io.Writer, fields []string, event *beat.Event) error {
    ...
}
```

btw. `[]byte(<string>)` has a high chance to allocate some temporary buffer and copy the string (as we pass the buffer to an interface). The `io.WriteString` function checks if the Writer passed implements the `io.StringWriter` interface and passes the original string instead of creating a copy.

urso commented on libbeat/processors/fingerprint/fingerprint.go:
```diff
@@ -0,0 +1,134 @@
+		}
+		if err != nil {
+			return "", errors.Wrapf(err, "failed when finding field [%v] in event", k)
+		}
```
Do we want to have an option to ignore missing fields in case we have at least one field present?

ycombinator replied:
Sorry, but is this the same as your suggestion in https://github.com/elastic/beats/pull/14205#discussion_r340863983 or something different?

urso replied:
almost. My suggestion originally was to add support to ignore missing fields. But apparently we can have other error types as well. Would it make sense to treat those other types as 'missing' as well?

ycombinator replied:
I see. So if we can't "get" a field for whatever reason, we treat it as missing and then, if the `missing_fields` option is set, ignore it. Hmm, I think this makes sense but let me just look into what other types of errors (besides `common.ErrKeyNotFound`) might be returned here.

ycombinator replied:
The only other error that can be returned here is if we try to get a value for a nested field (e.g. `a.b.c`), but the ancestor path to the field (`a.b`) does not resolve to a `map`. To me this also feels like a missing field as, even in this case, we still could not find `a.b.c` for the user. So I'm okay with collapsing the two error cases into one and adding `ignore_missing` handling to it.

urso commented on libbeat/processors/fingerprint/fingerprint_test.go:
```diff
@@ -0,0 +1,390 @@
+	tests := []struct {
+		name   string
+		fields []string
+	}{
```
common pattern: `tests := map[string]struct { fields []string }`. Using a map also randomized run order.

urso commented on libbeat/processors/fingerprint/fingerprint_test.go:
```diff
@@ -0,0 +1,390 @@
+
+	for _, test := range tests {
+		name := fmt.Sprintf("testing %v encoding", test.encoding)
+		t.Run(name, func(t *testing.T) {
```
Test names can be filtered using regexes. For this I like to give it some structure instead of having full names (assuming that the parent test its name is expressive enough). e.g. just pass encoding as name to `t.Run`. (In case I have multiple parameters or want a name I use `fmt.Sprintf("field=%v" , value)`).

In this case the test name could just be `TestEncoding/base64`. I can run the test using `go test -run Encoding/base64` (the `/` acts as a delimiter.

dedemorton commented on libbeat/docs/processors-using.asciidoc:
```diff
@@ -1184,6 +1184,28 @@ The following settings are supported:
+
+.Fingerprint options
+[options="header"]
+|======
```
You forgot the header row here. :-)  The rendered table looks like this: 

![image](https://user-images.githubusercontent.com/14206422/67908010-107e5a80-fb37-11e9-8431-148dd05ef2e4.png)

dedemorton commented on libbeat/docs/processors-using.asciidoc:
```diff
@@ -1184,6 +1184,28 @@ The following settings are supported:
+[options="header"]
+|======
+| `fields`          | yes       |               | List of fields to use as the source for the fingerprint.                                                               |
+| `target_field`    | no        | `fingerprint` | Field in which the generated fingerprint should be stored.                                                             |
```
Tables with more than 3 columns don't look very good in our HTML output (see screen capture in previous comment). Docbook swallows all table attributes, so we can't do anything about that right now. You could remove the example row and provide the example as part of the description. Or wait for some shift in the universe that will make this right.

ycombinator replied:
I replaced the table with a definition list, like the one used in the Extract Array processor, for instance. Let me know what you think. Thanks!

ycombinator commented:
@urso @dedemorton Thanks for your reviews. I believe I've addressed all your feedback now. This PR is ready for re-review, when you get a chance. Thanks again!

ycombinator commented:
jenkins, test this

urso commented on libbeat/processors/fingerprint/fingerprint.go:
```diff
@@ -0,0 +1,133 @@
+		i := v
+		switch vv := v.(type) {
+		case map[string]interface{}, []interface{}, common.MapStr:
+			return errors.Errorf("cannot compute fingerprint using non-scalar field [%v]", k)
```
would it make sense to apply 'IgnoreMissing' here as well?

ycombinator replied:
I'm not sure about this one. This case would be reached if the user specified a field that resolved to a non-scalar value. So the field would not actually be missing in this case.

I see this one as more of an error on the user's part with maybe making a mistake in specifying the field so I think we should tell the user about it via an error.

WDYT?

urso replied:
I'm fine with that.

urso commented on libbeat/processors/fingerprint/fingerprint.go:
```diff
@@ -0,0 +1,133 @@
+			i = vv.UTC()
+		}
+
+		io.WriteString(to, fmt.Sprintf("|%v|%v", k, i))
```
`fmt.Fprintf(to, "|%v|%v", k, i)`

urso commented on libbeat/processors/fingerprint/fingerprint.go:
```diff
@@ -0,0 +1,133 @@
+}
+
+func unique(in []string) []string {
+	seen := map[string]bool{}
```
a stringset can be implemented more space efficient via `map[string]struct{}{}`. The empty structs size is 0.

urso commented on libbeat/processors/fingerprint/fingerprint.go:
```diff
@@ -0,0 +1,133 @@
+	seen := map[string]bool{}
+	var out = make([]string, 0, len(in))
+	for _, item := range in {
+		if _, found := seen[item]; !found {
```
if a key is not known then the a lookup returns the zero value. The zero value of bool is `false`.

urso commented on libbeat/processors/fingerprint/fingerprint.go:
```diff
@@ -0,0 +1,133 @@
+			seen[item] = true
+			out = append(out, item)
+		}
+	}
```
as we build a map anyways :)

```
keys := make(map[string]struct{}, len(in))  // reserve some space when we create the map
for _, name := range in {
  keys[name] = struct{}{}
}
out := make([]string, 0, len(keys)) // only allocated the amount of memory we will really need
for name := range keys {
  out = append(out, name)
}
```

ycombinator commented:
@urso Thanks for the review again.

I found `common.StringSet` and decided to use that in place of the `unique` function I had implemented. However, I do appreciate your comments on it as I learnt a few things in the process!

Also, please see my comment about the `IgnoreMissing` question: https://github.com/elastic/beats/pull/14205#discussion_r341213621.

urso commented on libbeat/processors/fingerprint/fingerprint_test.go:
```diff
@@ -0,0 +1,385 @@
+				"fields": []string{},
+				"method": "sha256",
+			},
+			"failed to unpack fingerprint processor configuration: empty field accessing 'fields'",
```
nit; it's ok for now. But instead of checking strings one should consider typed errors that also carries some context.

ycombinator replied:
Implemented typed errors in 49a0e4f7286d22d2bcc121ac23fff06ce52edd09 and 65627787e402421c7657a53f0d7489792c3675f8.

ycombinator commented:
jenkins, test this

ycombinator commented:
jenkins, test this

ycombinator commented:
Travis CI is green. Jenkins CI is red because of `x-pack/agent` builds being yellow, which is unrelated to this PR. Merging.

neu5ron commented:
I am not sure where to move this issue forward, but it should be noted that the fingerprint processor for ingest processor creates inconsistent results for anybody else who may be using the same hashing techniques (md5, sha1, sha256, sha512, etc..)

neu5ron commented:
I am not sure where to move this issue forward, but it should be [noted that the fingerprint processor for ingest processor creates inconsistent](https://github.com/elastic/elasticsearch/issues/98339) results for anybody else who may be using the same hashing techniques (md5, sha1, sha256, sha512, etc..)
